### PR TITLE
Correcting binary_mask error; tests for mask; tweak to contributing.md

### PR DESCRIPTION
**Pull request by ns-rse**
This is the PR that #99 was meant to be and merges into `dev` the following changes...

* Correcting the mistake I'd made converting `binary_mask` to `boolean` when it should remain a `np.array`.
* Introduces tests for `masking` after having split it out into a function instead of a line in `topotracing.py`.
* Tweak to `contributions.md` to clarify from which branch people should branch from.

### CODE REVIEW

**Comment by ns-rse** [comment]
@bobturneruk apologies for the excessive commits and size of this Pull Request. This is the functional approach with separate modules. I'd like to do a quick comparison of this functionality with classes to see how they hang together and whether others would like to adopt this approach, will work on that over the next few days.

**Comment by bobturneruk** [comment]
Thanks for this @ns-rse - I'll take a look.

**Comment by bobturneruk** [comment]
We should perhaps have some install instructions + instructions to install `pytest` and run test? Or maybe a [dev] option in the setup?

**Comment by ns-rse** [comment]
> General stuff:
> 
> * We should perhaps have some install instructions + instructions to install `pytest` and run test? Or maybe a [dev] option in the setup?

Sure I can add an `INSTALL.md` which will document how to install from Git for now (with a view to packaging and releasing on PyPi further down the line.

Installing dependencies is part of what I've started doing with using [setuptools]() in its new form (its changed recently). My experience to date is to not install packages that are only required for tests, instead they are only pulled in when you explicitly want to test and are defined under...

```
[options.extras_require]
tests =
  pytest
  pytest-cov
  pytest-mpl
```

(used to be a section under `tests_requre = ` but that is now obsolete). If someone wants to explicitly run the tests I think its a case of specifying during installation with something like `pip install -e . test` and that would download and install the tests (albeit in the current directory rather than the virtual environment, see [thread if interested](https://stackoverflow.com/questions/4734292/specify-where-to-install-tests-require-dependencies-of-a-distribute-setuptools)).

I'll add how to do this into the `INSTALL.md`.

> * Could the tests for `filters.py` be organised such that that we have a `test_filters.py` that contains unit tests for the functions in `filters.py` that appear in the same order as the functions being tested?

I'll look at changing the order of tests in the `tests/test_filters.py` file to match that of the order they are in `topostats/filters.py`. It should already be close.

**Comment by ns-rse** [comment]
> Could the tests for filters.py be organised such that that we have a test_filters.py that contains unit tests for the functions in filters.py that appear in the same order as the functions being tested?

Calling the functions in order is covered in `tests/test_filters_minicircle.py` which runs through the steps of processing `minicircle.spm`. There are a few functions which aren't called directly when doing so. For example e.g. both `align_rows` and `remove_x_y_tilt` call the `row_col_quantiles` function, but this is never called directly in the processing itself. Similarly `remove_x_y_tilt` calls `calc_gradient` which in turn calls `calc_diff`.

In `test_filters.py` I set out to test the functions in isolation but under different circumstances (e.g. `row_col_quantiles()` with and without masking for both a small array (10x10) and the larger `image_random` fixture (which is 1024x1024).  I'd missed out tests for some of the functions not called directly which I've now added in, along with some linting corrections.

Its not that logical, but I hope the approach makes sense. 

In the refactoring I'm working on that uses a Class the methods not called directly are private. Opinions on whether to test private methods varies, some say only test what users see, I'm happy to have small/simple tests of private methods too though.  

Having a Class for this was my main aim, but felt getting tests in place based on functions first was useful (and it has been as the images serve as tests for the class as I develop it).

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:48]: Should this be in `filters.py`?

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:75]: Is this much different to `get_mask` in `filters.py`?

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:23]: Nothing in `numpy` for this?

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:135]: Does this add much to the function it's wrapping? Do we need it?

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:153]: Why wrap `label2rgb`?

File: topostats/find_grains.py
Comment (bobturneruk) [inline topostats/find_grains.py:170]: Why wrap `regionprops`?

**Review by bobturneruk — CHANGES_REQUESTED** [review]
Hi! I've at least glanced at all of this now, so no more feedback for the time being unless there are any changes. I think this can be merged once all this is addressed, but as discussed we need to check in with the team on:
- Use of `yaml` for config files
- Introduction of the command line interface

File: topostats/find_grains.py
Comment (ns-rse) [reply topostats/find_grains.py:75]: No its not and I did have a `FIXME` in there to reduce the duplication (which `pylint` complained about), but as I wanted to re-write these as classes I've taken the opportunity to abstract `get_mask()` to `utils.py` and import and use it in both the `Filters` and `FindGrains` classes I've been working on.

File: topostats/io.py
Comment (ns-rse) [reply topostats/io.py:6]: Contingency really, its what I've used in the past as others I worked with used it in preference to `pyyaml` on which `ruamel.yaml` is based.

A more useful answer is that it has better support for YAML 1.2,see [comparison](https://yaml.readthedocs.io/en/latest/pyyaml.html).

File: tests/test_io.py
Comment (ns-rse) [reply tests/test_io.py:12]: Ooh, that looks really neat, thank you for the pointers.

File: .github/workflows/tests.yaml
Comment (ns-rse) [reply .github/workflows/tests.yaml:28]: It's a useful plugin but its not very obvious how it works, you have to `return fig` at the end, then the first time you run with the option `pytest --mpl-generate-images=<path/to/save/images/to>` and it will `skip` tests with the `@pytest.mark.mpl_image_compare` fixture. When you then run the tests again without the option it compares the images generate by the test to those that existed.

Fairly circular chicken and egg situation, but its been useful as I move to refactoring to classes to use the images I generated with these functions to test I get the correct images with the classes.

File: config/example.yaml
Comment (ns-rse) [reply config/example.yaml:1]: Sure thing :+1:

File: setup.cfg
Comment (ns-rse) [reply setup.cfg:11]: It no doubt has an effect there (e.g. if the test matrix included Python 3.7 it would fail), but this is really in place for when users are installing to ensure they have Python `>= 3.8`.

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:20]: Sure that makes sense as it is an i/o operation.

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:41]: I'm wary of `utils` sub-modules becoming a catch all for everything to get put into. 

Utilities should, I feel, be things that are reused across other modules and as `extract_img_name()`, `extract_channel()` and `extract_pixels()` functions are only used as part of the process at the start of loading the image I opted to keep them here (I had in mind them being methods of the `Filters` class I've been working on when I wrote this all as functions, see branch [ns-rse/90-refactor-filters](https://github.com/AFM-SPM/TopoStats/tree/ns-rse/90-refactor-filters)).

Happy to move though, but let me know what you think about the above.

File: INSTALL.md
Comment (ns-rse) [reply INSTALL.md:1]: Sorry, it should have been a separate PR rather than bundling it in here. Would you like it moved?

File: topostats/logs/logs.py
Comment (ns-rse) [reply topostats/logs/logs.py:23]: Good idea, I've added an example to the docstring.

File: topostats/topotracing.py
Comment (ns-rse) [reply topostats/topotracing.py:20]: They don't conflict and only one set of messages is output, but I should have removed this and just used `LOGGER = logging.getLogger(LOGGER_NAME)` after I added it to `__init__.py`. Thanks for picking this up, now corrected.

File: topostats/topotracing.py
Comment (ns-rse) [reply topostats/topotracing.py:62]: Done, Emacs has the handy [numpydoc package](https://github.com/emacsmirror/numpydoc) that eases the task of writing docstrings based on the `def():`.

File: topostats/find_grains.py
Comment (ns-rse) [reply topostats/find_grains.py:48]: Based on its name probably but in the code I was working from it was under the grain finding section so I've stuck with that convention. Maybe its something @SylviaWhittle could direct us on.

File: topostats/find_grains.py
Comment (ns-rse) [reply topostats/find_grains.py:135]: For this and other functions I was intending to add `**kwargs` to these as I move from functions to classes and methods, allowing users to modify the calls to `label()`, `lab2rgb()`, `regionprops()`, which is why I wrap them up in a function.

At the moment I don't see much scope for people tweaking things to this level, but figured its useful to build this level of flexibility in from the outset.

File: tests/test_io.py
Comment (ns-rse) [reply tests/test_io.py:12]: Opted to use a more simple `test.yaml` that won't change as the config grows in order to validate `read_yaml()` works correctly using `TestCase().assertDictEqual()`.

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:261]: Moved threshold to `utils` although that in itself will change in light of #118 (although that will need a little more work once this PR is merged).

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:128]: The `row_col_quantiles()`, `calc_diff()` and `calc_gradient()` are only used in the filtering stage and to avoid `utils.py` growing unnecessarily figured its worth keeping them in the module in which they are used.

Happy to change though.

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:10]: I think this is a good idea and it should probably sit under `io`. It might perhaps be worthy of its own issue to address though so that an appropriate design pattern can be used to make the process flexible and open to extension to accommodate other file types.

File: topostats/filters.py
Comment (ns-rse) [reply topostats/filters.py:88]: As with above comment about moving `pySPM` related functions to their own module I think `extract_pixels()` and `extract_channel()` should move too (as they both call methods on the `SPM_image` objects).

File: tests/test_filters.py
Comment (ns-rse) [reply tests/test_filters.py:2]: Not sure, I put `disable=no-name-in-module` in there before I split the `find_grains.py` and related testing out as pylint was complaining, it doesn't complain now though.

The `disable=invalid-name` is only required in `test_filters_minicircle.py` though as the `dx` variable doesn't conform to `snake_case` but having chatted with @SylviaWhittle this morning the configuration option `dx` is no longer required as its obtained from the pySPM meta-data.

File: topostats/find_grains.py
Comment (ns-rse) [reply topostats/find_grains.py:2]: Yes, probably worthwhile, although we'll need to get consensus on linting across all contributors.

File: topostats/filters.py
Comment (bobturneruk) [reply topostats/filters.py:18]: @ns-rse - what you say is right. If they're only used here, this is where they should be.

File: topostats/filters.py
Comment (bobturneruk) [reply topostats/filters.py:10]: Please can you raise such an issue post merge @ns-rse?

File: topostats/filters.py
Comment (bobturneruk) [reply topostats/filters.py:81]: Cool - issue to move to io (not utils).

File: topostats/filters.py
Comment (bobturneruk) [reply topostats/filters.py:236]: Again, if only used here, keep here.

File: topostats/io.py
Comment (bobturneruk) [reply topostats/io.py:7]: Great, thanks.

File: topostats/logs/logs.py
Comment (bobturneruk) [reply topostats/logs/logs.py:23]: Looks good!

File: topostats/topotracing.py
Comment (bobturneruk) [reply topostats/topotracing.py:24]: Was discussed in meeting on 20220506 - YAML and CLI agreed.

File: topostats/find_grains.py
Comment (bobturneruk) [reply topostats/find_grains.py]: Let's leave it here for now.

File: tests/conftest.py
Comment (bobturneruk) [reply tests/conftest.py]: Looks like it now is.

**Pull request merged**
